### PR TITLE
electron: allow accessing the metrics endpoint for performance analysis

### PR DESCRIPTION
**Pull request by xai**
#### What it does

By default, when running Theia in Electron, all endpoints are protected by the ElectronTokenValidator.
This patch allows accessing the '/metrics' endpoint without a token, which enables us to collect metrics for performance analysis.

For this, ElectronTokenValidator is extended to allow access to the metrics endpoint. All other endpoints are still protected.

Contributed on behalf of STMicroelectronics

#### How to test

* Build and start the Electron example. To avoid getting a random free port on runtime, use `yarn electron start --port=3000`.
* Access the '/metrics' endpoint, e.g., by calling `curl http://localhost:3000/metrics`.
* Verify that metrics are returned.
* Access a different endpoint, e.g. `http://localhost:3000/foo`.
* Verify that the request is denied.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/metrics/src/electron-node/electron-metrics-backend-module.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import { ElectronTokenValidator } from '@theia/core/lib/electron-node/token/electron-token-validator';
+import { BackendApplicationContribution } from '@theia/core/lib/node';
+import { ExtensionMetricsContribution, MetricsBackendApplicationContribution, MetricsContribution, NodeMetricsContribution } from '../node';
+import { ConnectionHandler, RpcConnectionHandler, bindContributionProvider } from '@theia/core';
+import { measurementNotificationServicePath } from '../common';
+import { MeasurementMetricsBackendContribution } from '../node/measurement-metrics-contribution';
+
+export default new ContainerModule((bind, unbind, isBound, rebind) => {
+    bind(MetricsElectronTokenValidator).toSelf().inSingletonScope();
+    rebind(ElectronTokenValidator).to(MetricsElectronTokenValidator);
```
Comment (sdirix) [inline packages/metrics/src/electron-node/electron-metrics-backend-module.ts:23]: The only difference to the regular backend is the electron token validator rebinding. I would like to suggest moving this into a separate `ContainerModule` and in the `package.json`'s `theiaExtensions` add a new entry only containing the `backendElectron` entry.

This way we could avoid all of the duplicated binding code as the new Container would only contain the electron token validation specifics.

**Review by sdirix — COMMENTED** [review]
This works, however we should think about whether we really want to expose the `metrics` endpoint by default in Electron. I think it would be more reasonable to only do the rebinding in case we are in debug or development mode. We already only collect startup metrics in development mode, so we could align with that.

Rebinding the `ElectronTokenValidator` will also often not have an effect in downstream projects. It's a common pattern to rebind it to allow for custom exposed APIs, so they will overwrite this one again. We should probably have a contribution point within the `ElectronTokenValidator` to allow contributing whitelists. Of course this is out of scope for this PR, but we should create a ticket for that.

File: packages/metrics/src/electron-node/electron-metrics-backend-module.ts (hunk 1 above)
Comment (sdirix) [reply packages/metrics/src/electron-node/electron-metrics-backend-module.ts:23]: Already resolved :+1:

**Comment by xai** [comment]
Thanks, @sdirix and @tortmayr! I will revise my PR accordingly.

**Review by tortmayr — APPROVED** [review]
Looks good to me and works as expected.
One minor regarding imports but otherwise everything works.

**Comment by JonasHelming** [comment]
@tortmayr Can you merge this?

**Pull request merged**